models/src/node/mod.rs

```diff
@@ -0,0 +1,264 @@
+
+impl BottlerocketNodeState {
+    /// Returns the next state in the state machine if the current state has been reached successfully.
+    pub fn on_success(&self) -> Self {
+        match self {
+            Self::WaitingForUpdate => Self::PreparedToUpdate,
+            Self::PreparedToUpdate => Self::PerformedUpdate,
+            Self::PerformedUpdate => Self::RebootedToUpdate,
+            Self::RebootedToUpdate => Self::MonitoringUpdate,
+            Self::MonitoringUpdate => Self::WaitingForUpdate,
+        }
+    }
+
+    /// Returns the total time that a node can spend transitioning *from* the given state to the next state in the process.
+    pub fn timeout_time(&self) -> Option<Duration> {
+        match self {
+            Self::WaitingForUpdate => PREPARED_TO_UPDATE_TIMEOUT,
+            Self::PreparedToUpdate => PERFORMED_UPDATE_TIMEOUT,
+            Self::PerformedUpdate => REBOOTED_TO_UPDATE_TIMEOUT,
+            Self::RebootedToUpdate => MONITORING_UPDATE_TIMEOUT,
+            Self::MonitoringUpdate => WAITING_FOR_UPDATE_TIMEOUT,
+        }
+    }
+}
+
+// We can't use these consts inside macros, but we do provide constants for use in generating kubernetes objects.
+pub const K8S_NODE_KIND: &str = "BottlerocketNode";
+pub const K8S_NODE_PLURAL: &str = "bottlerocketnodes";
+pub const K8S_NODE_STATUS: &str = "bottlerocketnodes/status";
+pub const K8S_NODE_SHORTNAME: &str = "brn";
+
+/// The `BottlerocketNodeSpec` can be used to drive a node through the update state machine. A node
+/// linearly drives towards the desired state. The brupop controller updates the spec to specify a node's desired state,
+/// and the host agent drives state changes forward and updates the `BottlerocketNodeStatus`.
+#[derive(
+    Clone,
+    CustomResource,
+    Serialize,
+    Deserialize,
+    Debug,
+    Default,
+    Eq,
+    PartialEq,
+    JsonSchema,
+    Validate,
+)]
+#[kube(
+    derive = "Default",
+    derive = "PartialEq",
+    group = "brupop.bottlerocket.aws",
+    kind = "BottlerocketNode",
+    namespaced,
+    plural = "bottlerocketnodes",
+    shortname = "brn",
+    singular = "bottlerocketnode",
+    status = "BottlerocketNodeStatus",
+    version = "v1",
+    printcolumn = r#"{"name":"State", "type":"string", "jsonPath":".status.current_state"}"#,
+    printcolumn = r#"{"name":"Version", "type":"string", "jsonPath":".status.current_version"}"#,
+    printcolumn = r#"{"name":"Target State", "type":"string", "jsonPath":".spec.state"}"#,
+    printcolumn = r#"{"name":"Target Version", "type":"string", "jsonPath":".spec.version"}"#
+)]
+pub struct BottlerocketNodeSpec {
+    /// Records the desired state of the `BottlerocketNode`
+    pub state: BottlerocketNodeState,
+    /// The time at which the most recent state was set as the desired state.
+    state_transition_timestamp: Option<String>,
+    /// The desired update version, if any.
+    #[validate(regex = "SEMVER_RE")]
+    version: Option<String>,
+}
+
+impl BottlerocketNode {
```

Comment:
Where is this coming from? Is it a struct? But I couldn't find the definition for this struct.

Reply:
This struct is generated when deriving `CustomResource` and the `kube` macro on the `BottlerocketNodeSpec` struct. See [`CustomResource`](https://docs.rs/kube-derive/0.64.0/kube_derive/derive.CustomResource.html) documentation.